domain/modelagent/modelmigration/export.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+
+	"github.com/juju/description/v9"
+
+	coreagentbinary "github.com/juju/juju/core/agentbinary"
+	"github.com/juju/juju/core/logger"
+	coremachine "github.com/juju/juju/core/machine"
+	"github.com/juju/juju/core/modelmigration"
+	coreostype "github.com/juju/juju/core/os/ostype"
+	"github.com/juju/juju/core/semversion"
+	coreunit "github.com/juju/juju/core/unit"
+	modelagentservice "github.com/juju/juju/domain/modelagent/service"
+	modelagentstate "github.com/juju/juju/domain/modelagent/state"
+	"github.com/juju/juju/internal/errors"
+)
+
+// baseAgentBinaryExportOperation describes the base set of operation
+// characteristics shared between export operations of this package.
+// Specifically the common need for the export service.
+type baseAgentBinaryExportOperation struct {
+	modelmigration.BaseOperation
+	exportService ExportService
+	logger        logger.Logger
+}
+
+// ExportService describes the service required for exporting agent binary
+// versions of a model to a new controller.
+type ExportService interface {
+	GetMachineReportedAgentVersion(
+		ctx context.Context,
+		machineName coremachine.Name,
+	) (coreagentbinary.Version, error)
+
+	GetUnitReportedAgentVersion(
+		ctx context.Context,
+		unitName coreunit.Name,
+	) (coreagentbinary.Version, error)
+}
+
+// exportMachineAgentBinaryOperation is a model migration export operation for
+// setting the reported agent binary version for each machine being exported as
+// part of a model.
+type exportMachineAgentBinaryOperation struct {
+	baseAgentBinaryExportOperation
+}
+
+// exportUnitAgentBinaryOperation is a model migration export operation for
+// setting the reported agent binary version for each machine being exported as
+// part of a model.
+type exportUnitAgentBinaryOperation struct {
+	baseAgentBinaryExportOperation
+}
+
+// Execute will attempt to set the reported tools/agentbinary version for each
+// machine in the model description. This operation assumes that prior to this
+// being performed all of the machines in the model have been exported on to
+// the model description.
+func (e *exportMachineAgentBinaryOperation) Execute(
+	ctx context.Context,
+	model description.Model,
+) error {
+	for _, machine := range model.Machines() {
+		mName := coremachine.Name(machine.Id())
+		agentVersion, err := e.exportService.GetMachineReportedAgentVersion(ctx, mName)
+
+		if err != nil {
+			return errors.Errorf(
+				"getting reported agent version for machine %q: %w",
+				mName, err,
+			)
+		}
+
+		exportVer := semversion.Binary{
+			Number:  agentVersion.Number,
+			Arch:    agentVersion.Arch,
+			Release: coreostype.Ubuntu.String(),
```

Comment:
Are we going to stop using release?

Reply:
Yes we are just defaulting it off to Ubuntu where we still have to keep using `Binary` still. Unfortunately `Binary` has hooks into params and the client so this is no easy task.

Anywhere we have to use it at the moment we can just default the value.